### PR TITLE
remove CompleteTensorType

### PR DESCRIPTION
**Pull request by zdevito**
This replaces uses of CompleteTensorType with ProfiledTensorType. 

Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#24169 remove CompleteTensorType**
* #24077 Remove DimensionedTensorType

Differential Revision: [D16765329](https://our.internmc.facebook.com/intern/diff/D16765329)

### CODE REVIEW

File: torch/csrc/jit/fuser/tensor_desc.h [hunk 1]
```diff
@@ -41,8 +41,11 @@ struct TORCH_API TensorDesc {
   TensorDesc(const at::Tensor& t)
       : TensorDesc(t.scalar_type(), t.sizes(), t.strides()) {}
 
-  TensorDesc(const c10::CompleteTensorTypePtr& type)
-      : TensorDesc(type->scalarType(), type->sizes(), type->strides()) {}
+  TensorDesc(const c10::ProfiledTensorTypePtr& type)
+      : TensorDesc(
```
Comment (Krovatkin) [inline torch/csrc/jit/fuser/tensor_desc.h]: nitpick: probably won't hurt to stick in `INTERNAL_TORCH_ASSERT(type->isComplete())` in places like this

File: aten/src/ATen/core/jit_type.h [hunk 2]
```diff
@@ -658,6 +622,10 @@ struct CAFFE2_API ProfiledTensorType : public TensorType {
     auto gr = merge_primitive(requiresGrad(), other->requiresGrad());
     return ProfiledTensorType::create(scalar_type, dev, sz, srs, gr);
   }
+  // is all information about the type specified except for autograd?
```
Comment (Krovatkin) [inline aten/src/ATen/core/jit_type.h]: maybe we could a quick comment as to why `requires_gard` is excluded?

File: torch/csrc/jit/passes/peephole.cpp [hunk 3]
```diff
@@ -46,9 +46,11 @@ void PeepholeOptimizeImpl(Block* block, bool addmm_fusion_enabled) {
       // x.expand(x.size()) == x
       if (auto input_type = node->namedInput(attr::self)
                                 ->type()
-                                ->cast<CompleteTensorType>()) {
+                                ->cast<ProfiledTensorType>()) {
         auto expanded_sizes = node->get<c10::List<int64_t>>(attr::size);
-        if (!expanded_sizes.has_value() || c10::impl::toVector(*expanded_sizes) == input_type->sizes()) {
+        auto input_type_sizes = input_type->sizes().concrete_sizes();
+        if (expanded_sizes.has_value() && input_type_sizes &&
```
Comment (Krovatkin) [inline torch/csrc/jit/passes/peephole.cpp]: is omitting `!` in front of `expanded_sizes.has_value()` intentional?

**Review by Krovatkin — APPROVED** [review]
:shipit: a few minor comments

File: test/test_jit.py [hunk 4]
```diff
@@ -8209,9 +8209,9 @@ def tensor_unifying(x, y, z):
 
         graph = _propagate_shapes(tensor_unifying.graph, (a, b, c), False)
         if_outputs = list(graph.findNode("prim::If").outputs())
-        self.assertTrue(if_outputs[0].type().str() == "Float(2, 2)")
-        self.assertTrue(if_outputs[1].type().str() == "Tensor(2, *)")
-        self.assertTrue(if_outputs[2].type().str() == "Tensor(2, 4)")
+        self.assertTrue(if_outputs[0].type().str() == "Float(*, *)")
```
[the remark this answers is not in the document]
Comment (zdevito) [reply test/test_jit.py]: I changed this back to what it was before 'remove DimensionedTensorType', now propagate shapes only sets the dimensions and not the sizes. There were some tests that relied on this behavior.

File: torch/csrc/jit/passes/peephole.cpp (hunk 3 above)
Comment (zdevito) [reply torch/csrc/jit/passes/peephole.cpp]: yes, my reading was that this code was completely bogus before this change. Before it said 'if expanded_sizes is not a constant, then it is safe to remove the expand call'. Now it says 'if expanded_sizes is a constant, and  its value is the same as the input sizes, we can remove the expand'.

**Pull request closed without merging**